udev: Extract RAM properties from DMI information

hadess opened the pull request:
Add memory_id program to set properties about the physical memory
devices in the system. This is useful on machines with removable memory
modules to show how the machine can be upgraded, and on all devices to
detect the actual RAM size, without relying on the OS accessible amount.

Closes: #16651

@felipeborges, I'd appreciate a test! :)

This is the output of the helper (so the properties that should be set on the DMI device):
```
MEMORY_ARRAY_LOCATION=System Board Or Motherboard
MEMORY_ARRAY_MAX_CAPACITY=68719476736
MEMORY_ARRAY_NUM_DEVICES=4
MEMORY_DEVICE_TOTAL_WIDTH0=64
MEMORY_DEVICE_DATA_WIDTH0=64
MEMORY_DEVICE_SIZE0=8589934592
MEMORY_DEVICE_FORM_FACTOR0=DIMM
MEMORY_DEVICE_LOCATOR0=ChannelA-DIMM0
MEMORY_DEVICE_BANK_LOCATOR0=BANK 0
MEMORY_DEVICE_TYPE0=DDR4
MEMORY_DEVICE_TYPE_DETAIL0=Synchronous
MEMORY_DEVICE_SPEED_MTS0=2667
MEMORY_DEVICE_MANUFACTURER0=Samsung
MEMORY_DEVICE_SERIAL_NUMBER0=416433E9
MEMORY_DEVICE_ASSET_TAG0=9876543210
MEMORY_DEVICE_PART_NUMBER0=M378A1K43CB2-CTD    
MEMORY_DEVICE_RANK0=1
MEMORY_DEVICE_CONFIGURED_SPEED_MTS0=2400
MEMORY_DEVICE_MINIMUM_VOLTAGE0=1.2
MEMORY_DEVICE_MAXIMUM_VOLTAGE0=1.2
MEMORY_DEVICE_CONFIGURED_VOLTAGE0=1.2
MEMORY_DEVICE_TOTAL_WIDTH1=64
MEMORY_DEVICE_DATA_WIDTH1=64
MEMORY_DEVICE_SIZE1=8589934592
MEMORY_DEVICE_FORM_FACTOR1=DIMM
MEMORY_DEVICE_LOCATOR1=ChannelA-DIMM1
MEMORY_DEVICE_BANK_LOCATOR1=BANK 1
MEMORY_DEVICE_TYPE1=DDR4
MEMORY_DEVICE_TYPE_DETAIL1=Synchronous
MEMORY_DEVICE_SPEED_MTS1=2400
MEMORY_DEVICE_MANUFACTURER1=859B
MEMORY_DEVICE_SERIAL_NUMBER1=A02550A6
MEMORY_DEVICE_ASSET_TAG1=9876543210
MEMORY_DEVICE_PART_NUMBER1=BLT8G4D26AFTA.16FBD 
MEMORY_DEVICE_RANK1=2
MEMORY_DEVICE_CONFIGURED_SPEED_MTS1=2400
MEMORY_DEVICE_MINIMUM_VOLTAGE1=1.2
MEMORY_DEVICE_MAXIMUM_VOLTAGE1=1.2
MEMORY_DEVICE_CONFIGURED_VOLTAGE1=1.2
MEMORY_DEVICE_PRESENT2=0
MEMORY_DEVICE_LOCATOR2=ChannelB-DIMM0
MEMORY_DEVICE_BANK_LOCATOR2=BANK 2
MEMORY_DEVICE_TYPE2=Unknown
MEMORY_DEVICE_TYPE_DETAIL2=None
MEMORY_DEVICE_MANUFACTURER2=Not Specified
MEMORY_DEVICE_SERIAL_NUMBER2=Not Specified
MEMORY_DEVICE_ASSET_TAG2=Not Specified
MEMORY_DEVICE_PART_NUMBER2=Not Specified
MEMORY_DEVICE_TOTAL_WIDTH3=64
MEMORY_DEVICE_DATA_WIDTH3=64
MEMORY_DEVICE_SIZE3=8589934592
MEMORY_DEVICE_FORM_FACTOR3=DIMM
MEMORY_DEVICE_LOCATOR3=ChannelB-DIMM1
MEMORY_DEVICE_BANK_LOCATOR3=BANK 3
MEMORY_DEVICE_TYPE3=DDR4
MEMORY_DEVICE_TYPE_DETAIL3=Synchronous
MEMORY_DEVICE_SPEED_MTS3=2400
MEMORY_DEVICE_MANUFACTURER3=859B
MEMORY_DEVICE_SERIAL_NUMBER3=A0254F38
MEMORY_DEVICE_ASSET_TAG3=9876543210
MEMORY_DEVICE_PART_NUMBER3=BLT8G4D26AFTA.16FBD 
MEMORY_DEVICE_RANK3=2
MEMORY_DEVICE_CONFIGURED_SPEED_MTS3=2400
MEMORY_DEVICE_MINIMUM_VOLTAGE3=1.2
MEMORY_DEVICE_MAXIMUM_VOLTAGE3=1.2
MEMORY_DEVICE_CONFIGURED_VOLTAGE3=1.2
```

It supports using dmidecode generated dumps, for testing purposes.

The overall patch needs the udev rule actually tested, and probably more coding style fixes. I also don't find the property names particularly readable for the RAM modules, ideas welcome.

felipeborges commented:
This is the output in my machine https://paste.centos.org/view/f826e1ef

hadess commented:
> This is the output in my machine https://paste.centos.org/view/f826e1ef

16GB of soldered RAM :)

I don't like this though:
```
MEMORY_DEVICE_MINIMUM_VOLTAGE1=1.25
MEMORY_DEVICE_MAXIMUM_VOLTAGE1=1.25
MEMORY_DEVICE_CONFIGURED_VOLTAGE1=1.2
```

keszybz commented:
Hmm, wouldn't it be nicer to move the index before the last property? It'll be easier to grep and oggle by eye. (Also, if there's more than 10 devices, MEMORY_DEVICE_TOTAL_WIDTH1 and MEMORY_DEVICE_TOTAL_WIDTH10 would be even more annoying to grep.)

`MEMORY_DEVICE_TOTAL_WIDTH1` → `MEMORY_DEVICE_1_TOTAL_WIDTH` ?

hadess commented:
> I don't like this though:
> 
> ```
> MEMORY_DEVICE_MINIMUM_VOLTAGE1=1.25
> MEMORY_DEVICE_MAXIMUM_VOLTAGE1=1.25
> MEMORY_DEVICE_CONFIGURED_VOLTAGE1=1.2
> ```

It's what the firmware advertises, this is the output of dmidecode on the same system:
```
        Minimum Voltage: 1.25 V
        Maximum Voltage: 1.25 V
        Configured Voltage: 1.2 V
```

keszybz commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,777 @@
+/*
```
Where is this code coming from? Is GPLv2+ the only license that we can use? LGPLv2+ would be preferable to match the rest of the code base. Either way, please add a SPDX header.

hadess replied:
The scary bits with the hex numbers are from dmidecode, so GPLv2+.

hadess commented:
> `MEMORY_DEVICE_TOTAL_WIDTH1` → `MEMORY_DEVICE_1_TOTAL_WIDTH` ?

Sound good, let's fix this up.

keszybz commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,777 @@
+        else if (code & 0xFFC00UL)
+                capacity = code * 1024^3;
+        else
+                capacity = code * 1024^4;
```
That's very strange code. Was that supposed to be `1024, 1024*1024, 1024*1024*1024`?

hadess replied:
Yeah, the original code is:
https://github.com/mirror/dmidecode/blob/master/dmidecode.c#L2565-L2579

Completely messed up that one...

keszybz reviewed (commented):
Here I get:
<details>

```
MEMORY_ARRAY_LOCATION=System Board Or Motherboard
MEMORY_ARRAY_MAX_CAPACITY=21474836480
MEMORY_ARRAY_NUM_DEVICES=2
MEMORY_DEVICE_TOTAL_WIDTH0=64
MEMORY_DEVICE_DATA_WIDTH0=64
MEMORY_DEVICE_SIZE0=4294967296
MEMORY_DEVICE_FORM_FACTOR0=Chip
MEMORY_DEVICE_LOCATOR0=ChannelA
MEMORY_DEVICE_BANK_LOCATOR0=BANK 0
MEMORY_DEVICE_TYPE0=LPDDR3
MEMORY_DEVICE_TYPE_DETAIL0=Synchronous
MEMORY_DEVICE_SPEED_MTS0=1867
MEMORY_DEVICE_MANUFACTURER0=Samsung
MEMORY_DEVICE_SERIAL_NUMBER0=None
MEMORY_DEVICE_ASSET_TAG0=None
MEMORY_DEVICE_PART_NUMBER0=K4E6E304EE-EGCF   
MEMORY_DEVICE_RANK0=2
MEMORY_DEVICE_CONFIGURED_SPEED_MTS0=1867
MEMORY_DEVICE_CONFIGURED_VOLTAGE0=1.2
MEMORY_DEVICE_TOTAL_WIDTH1=64
MEMORY_DEVICE_DATA_WIDTH1=64
MEMORY_DEVICE_SIZE1=4294967296
MEMORY_DEVICE_FORM_FACTOR1=SODIMM
MEMORY_DEVICE_LOCATOR1=ChannelB
MEMORY_DEVICE_BANK_LOCATOR1=BANK 2
MEMORY_DEVICE_TYPE1=LPDDR3
MEMORY_DEVICE_TYPE_DETAIL1=Synchronous
MEMORY_DEVICE_SPEED_MTS1=1867
MEMORY_DEVICE_MANUFACTURER1=Samsung
MEMORY_DEVICE_SERIAL_NUMBER1=None
MEMORY_DEVICE_ASSET_TAG1=None
MEMORY_DEVICE_PART_NUMBER1=K4E6E304EE-EGCF   
MEMORY_DEVICE_RANK1=2
MEMORY_DEVICE_CONFIGURED_SPEED_MTS1=1867
MEMORY_DEVICE_CONFIGURED_VOLTAGE1=1.2
```
</details>

and on another machine
<details>

```
MEMORY_ARRAY_LOCATION=System Board Or Motherboard
MEMORY_ARRAY_EC_TYPE=Multi-bit ECC
MEMORY_ARRAY_MAX_CAPACITY=12884901888
MEMORY_ARRAY_NUM_DEVICES=3
MEMORY_ARRAY_LOCATION=System Board Or Motherboard
MEMORY_ARRAY_EC_TYPE=Multi-bit ECC
MEMORY_ARRAY_MAX_CAPACITY=12884901888
MEMORY_ARRAY_NUM_DEVICES=3
MEMORY_DEVICE_TOTAL_WIDTH0=72
MEMORY_DEVICE_DATA_WIDTH0=64
MEMORY_DEVICE_SIZE0=8589934592
MEMORY_DEVICE_FORM_FACTOR0=DIMM
MEMORY_DEVICE_LOCATOR0=CPU0 DIMM1
MEMORY_DEVICE_BANK_LOCATOR0=Not Specified
MEMORY_DEVICE_TYPE0=DDR3
MEMORY_DEVICE_TYPE_DETAIL0=Synchronous Unbuffered (Unregistered)
MEMORY_DEVICE_SPEED_MTS0=1600
MEMORY_DEVICE_MANUFACTURER0=JEDEC ID:80 AD
MEMORY_DEVICE_SERIAL_NUMBER0=BC8A0E80
MEMORY_DEVICE_ASSET_TAG0=Not Specified
MEMORY_DEVICE_PART_NUMBER0=HMT41GU7AFR8C-PB
MEMORY_DEVICE_RANK0=2
MEMORY_DEVICE_TOTAL_WIDTH1=72
MEMORY_DEVICE_DATA_WIDTH1=64
MEMORY_DEVICE_SIZE1=8589934592
MEMORY_DEVICE_FORM_FACTOR1=DIMM
MEMORY_DEVICE_LOCATOR1=CPU0 DIMM2
MEMORY_DEVICE_BANK_LOCATOR1=Not Specified
MEMORY_DEVICE_TYPE1=DDR3
MEMORY_DEVICE_TYPE_DETAIL1=Synchronous Unbuffered (Unregistered)
MEMORY_DEVICE_SPEED_MTS1=1600
MEMORY_DEVICE_MANUFACTURER1=JEDEC ID:80 AD
MEMORY_DEVICE_SERIAL_NUMBER1=86A6934F
MEMORY_DEVICE_ASSET_TAG1=Not Specified
MEMORY_DEVICE_PART_NUMBER1=HMT41GU7AFR8C-PB
MEMORY_DEVICE_RANK1=2
MEMORY_DEVICE_DATA_WIDTH2=64
MEMORY_DEVICE_PRESENT2=0
MEMORY_DEVICE_FORM_FACTOR2=DIMM
MEMORY_DEVICE_LOCATOR2=CPU0 DIMM3
MEMORY_DEVICE_BANK_LOCATOR2=Not Specified
MEMORY_DEVICE_TYPE2=DDR3
MEMORY_DEVICE_TYPE_DETAIL2=Synchronous
MEMORY_DEVICE_MANUFACTURER2=JEDEC ID:
MEMORY_DEVICE_SERIAL_NUMBER2= 
MEMORY_DEVICE_ASSET_TAG2=Not Specified
MEMORY_DEVICE_PART_NUMBER2= 
MEMORY_DEVICE_TOTAL_WIDTH3=72
MEMORY_DEVICE_DATA_WIDTH3=64
MEMORY_DEVICE_SIZE3=8589934592
MEMORY_DEVICE_FORM_FACTOR3=DIMM
MEMORY_DEVICE_LOCATOR3=CPU1 DIMM1
MEMORY_DEVICE_BANK_LOCATOR3=Not Specified
MEMORY_DEVICE_TYPE3=DDR3
MEMORY_DEVICE_TYPE_DETAIL3=Synchronous Unbuffered (Unregistered)
MEMORY_DEVICE_SPEED_MTS3=1600
MEMORY_DEVICE_MANUFACTURER3=JEDEC ID:80 AD
MEMORY_DEVICE_SERIAL_NUMBER3=A18A0E80
MEMORY_DEVICE_ASSET_TAG3=Not Specified
MEMORY_DEVICE_PART_NUMBER3=HMT41GU7AFR8C-PB
MEMORY_DEVICE_RANK3=2
MEMORY_DEVICE_TOTAL_WIDTH4=72
MEMORY_DEVICE_DATA_WIDTH4=64
MEMORY_DEVICE_SIZE4=8589934592
MEMORY_DEVICE_FORM_FACTOR4=DIMM
MEMORY_DEVICE_LOCATOR4=CPU1 DIMM2
MEMORY_DEVICE_BANK_LOCATOR4=Not Specified
MEMORY_DEVICE_TYPE4=DDR3
MEMORY_DEVICE_TYPE_DETAIL4=Synchronous Unbuffered (Unregistered)
MEMORY_DEVICE_SPEED_MTS4=1600
MEMORY_DEVICE_MANUFACTURER4=JEDEC ID:80 AD
MEMORY_DEVICE_SERIAL_NUMBER4=7D8A0E80
MEMORY_DEVICE_ASSET_TAG4=Not Specified
MEMORY_DEVICE_PART_NUMBER4=HMT41GU7AFR8C-PB
MEMORY_DEVICE_RANK4=2
MEMORY_DEVICE_DATA_WIDTH5=64
MEMORY_DEVICE_PRESENT5=0
MEMORY_DEVICE_FORM_FACTOR5=DIMM
MEMORY_DEVICE_LOCATOR5=CPU1 DIMM3
MEMORY_DEVICE_BANK_LOCATOR5=Not Specified
MEMORY_DEVICE_TYPE5=DDR3
MEMORY_DEVICE_TYPE_DETAIL5=Synchronous
MEMORY_DEVICE_MANUFACTURER5=JEDEC ID:
MEMORY_DEVICE_SERIAL_NUMBER5= 
MEMORY_DEVICE_ASSET_TAG5=Not Specified
MEMORY_DEVICE_PART_NUMBER5= 
MEMORY_DEVICE_TOTAL_WIDTH6=2
MEMORY_DEVICE_DATA_WIDTH6=2
MEMORY_DEVICE_SIZE6=2097152
MEMORY_DEVICE_FORM_FACTOR6=Chip
MEMORY_DEVICE_LOCATOR6=SYSTEM ROM
MEMORY_DEVICE_BANK_LOCATOR6=Not Specified
MEMORY_DEVICE_TYPE6=Flash
MEMORY_DEVICE_TYPE_DETAIL6=Non-Volatile
MEMORY_DEVICE_MANUFACTURER6=Not Specified
MEMORY_DEVICE_SERIAL_NUMBER6=Not Specified
MEMORY_DEVICE_ASSET_TAG6=Not Specified
MEMORY_DEVICE_PART_NUMBER6=Not Specified
```
</details>

keszybz commented on rules.d/70-memory.rules:
```diff
@@ -0,0 +1,8 @@
+# do not edit this file, it will be overwritten on update
+
+ACTION=="remove", GOTO="memory_end"
```
Hmm, "remove|unbind" ?

poettering replied:
"unbind" shouldn't happen for DMI devices. And even if, I think the rules should run too and update the data given what is available then. i.e. if the "unbind" means some attrs vanish, then great, the rules will try to read them, notice they don#t exist anymore, and handle that gracefully and not generate props.

i'd always write stuff gracefully. Unless we know for sure the whole thing is gone (i.e. on "remove") let's run as much as we can, but make sure to write the rules reasonably graceful.

keszybz commented on src/udev/memory_id/dmidecode-config.h:
```diff
@@ -0,0 +1,11 @@
+/* Use memory alignment workaround or not */
+#ifdef __ia64__
+#define ALIGNMENT_WORKAROUND
+#endif
```
Is there any need to keep this in a header? It's only used in `dmidecode.c` afaict.

hadess replied:
There's really no need for rhetorical questions about the coding style. Nearly everything you see comes from dmidecode, which very much looks like 1999 code. Just ask me to make the changes, and I'll do that, although I'm still waiting for feedback on whether the approach is the correct one...

keszybz commented on src/udev/memory_id/types.h:
```diff
@@ -0,0 +1,63 @@
+#define WORD(x) (uint16_t)(*(const uint16_t *)(x))
+#define DWORD(x) (uint32_t)(*(const uint32_t *)(x))
+#define QWORD(x) (*(const u64 *)(x))
+#endif /* ALIGNMENT_WORKAROUND || BIGENDIAN */
```
All this ugliness can be pushed in dmidecode.c, it's only used there.

keszybz reviewed (commented):
I don't want to suggest changes to the code before the general direction is accepted. @yuwata, @poettering please give at least your general review of the approach.

poettering commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,768 @@
+#include "types.h"
+
+#define out_of_spec "<OUT OF SPEC>"
+static const char *bad_index = "<BAD INDEX>";
```
hmm, why one as a macro and te other as global variable? weird? and even a ptr instread of an array?

poettering commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,768 @@
+ */
+
+static char *_dmi_string(const struct dmi_header *dm, uint8_t s)
+{
```
coding style, please `{` on same line as function name. See

https://systemd.io/CODING_STYLE/#formatting

poettering commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,768 @@
+
+static char *_dmi_string(const struct dmi_header *dm, uint8_t s)
+{
+        char *bp = (char *)dm->data;
```
should be const?

poettering commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,768 @@
+
+        bp += dm->length;
+        while (s > 1 && *bp)
+        {
```
bracket on same line as while, please

poettering commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,768 @@
+                return "Not Specified";
+
+        bp = _dmi_string(dm, s);
+        if (bp == NULL)
```
doesn't matter, but for ptr != NULL checks we usually rely on C's downgrade-to-bool feature, i.e. we'd write this as `if (!bp)…`

we insist on explicit comparisons only for variables that actually arithmetic inetegres or floats, i.e. not pointers, bools or boolean-like.

poettering commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,768 @@
+
+        capacity = code.h;
+        capacity = capacity << 32 | code.l;
+        if (shift == 1)
```
shift is a bool? should have type `bool` then?

poettering commented on src/udev/memory_id/types.h:
```diff
@@ -0,0 +1,63 @@
+typedef struct {
+        uint32_t h;
+        uint32_t l;
+} u64;
```
the kernel has a type called "u64", and it's not really defined like this. i'd rather we name this differently to avoid confusion. At least to me it's a first time I see a type called "u64" defined as struct...

hadess replied:
Called it `ULargeInteger`, inspired by https://docs.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-ularge_integer-r1

poettering commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,768 @@
+                "MCA Add-on Card",
+                "PCMCIA Add-on Card",
+                "Proprietary Add-on Card",
+                "NuBus" /* 0x0A */
```
maybe use the `[…] = ` notation for arrays like this? i.e.

static const char *location[] = {
        [0x01] = "Other",
…
```

And yes, this would mean we'd waste the first entry, but I think that's OK.

poettering commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,768 @@
+                "CXL Flexbus 1.0" /* 0xA4 */
+        };
+
+        if (code >= 0x01 && code <= 0x0A)
```
use ELEMENTSOF()?

hadess replied:
I can use ELEMENTSOF for the upper-bound, but not the lower.

poettering commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,768 @@
+ * 7.18 Memory Device (Type 17)
+ */
+
+static void dmi_memory_device_width(const char *attr_prefix, const char *attr_suffix, unsigned int slot_num, uint16_t code)
```
doesn't matter, but we usually just write "unsigned" rather than "unsigned int"

poettering commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,768 @@
+                return;
+        } else if (code == 0xFFFF) {
+                return;
+        } else {
```
doesn't matter, but we usually don't place {} around single-line if blocks

hadess replied:
I find it looks exceedingly weird when one conditional has them, but not the others. Let me know if you still want that after checking the last coding style changes.

poettering commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,768 @@
+{
+        uint64_t capacity;
+
+        capacity = code * 1024 * 1024;
```
hmm, you store the result in a 64bit value, but all factors are 32bit and of different sign. maybe write as:

```
capacity = (uint64_t) code * 1024U * 1024U;
```

poettering commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,768 @@
+        if (code == 0)
+                return;
+        if (code % 100)
+                printf("%s_%u_%s=%g\n", attr_prefix, slot_num, attr_suffix, (float)code / 1000);
```
"float" is such a weird type. We should never use that unless in an array where the memory savings matter. for all other cases use "double" please. In paticular as "%g" actually outputs a double, not a float anyway.

poettering commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,768 @@
+{
+        if (code == 0)
+                return;
+        if (code % 100)
```
we usually want explicit numeric comparisons for everything that isn't a pointer, a bool or boolean-like. i.e. please write as:

```c
if (code % 100 != 0)
…
```

poettering commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,768 @@
+        if (code % 100)
+                printf("%s_%u_%s=%g\n", attr_prefix, slot_num, attr_suffix, (float)code / 1000);
+        else
+                printf("%s_%u_%s=%.1f\n", attr_prefix, slot_num, attr_suffix, (float)code / 1000);
```
same here... %f actually outputs doubles, not floats, yet you explicitly cast to float here...

poettering commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,768 @@
+        if ((code & 0xFFFE) == 0)
+                printf("MEMORY_DEVICE_%u_TYPE_DETAIL=%s\n", slot_num, "None");
+        else
+        {
```
coding style, please put `{` on same line as `else`

poettering commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,768 @@
+                for (i = 1; i <= 15; i++)
+                        if (code & (1 << i))
+                                off += sprintf(list + off, off ? " %s" : "%s",
+                                               detail[i - 1]);
```
btw, why do you append everything in a buffer first and then output it with one printf()? you could just output things directly with printf() or fputs() or so. After all stdio is buffered io and internally maintains such buffers anyway. Would be a lot less fragile that way, given the manually calulcated array sizes won't be necessary anymore then.

poettering commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,768 @@
+         * Note: DMI types 37 and 42 are untested
+         */
+        switch (h->type)
+        {
```
coding style: please place { one same line as switch

poettering commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,768 @@
+        {
+        case 16: /* 7.17 Physical Memory Array */
+                log_debug("Physical Memory Array");
+                if (h->length < 0x0F) break;
```
coding style, please adding line break before `break`

poettering commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,768 @@
+        if (!found)
+                ret = EXIT_FAILURE;
+
+        free(buf);
```
use `_cleanup_free_` on the variable

poettering commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,768 @@
+        len = size;
+        dmi_table_decode(buf, len, num, flags);
+
+        free(buf);
```
use `_cleanup_free_`, i.e. gcc-style automatic cleanup function stuff

poettering commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,768 @@
+        h->type = data[0];
+        h->length = data[1];
+        h->handle = WORD(data + 2);
+        h->data = data;
```
use structured initialization, i.e. 

```
*h = (struct dmi_header) {
        .type = data[0],
        …
};
```

poettering commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,768 @@
+
+                /* Look for the next handle */
+                next = data + h.length;
+                while ((unsigned long)(next - buf + 1) < len
```
should be size_t, no?

poettering commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,768 @@
+                next += 2;
+
+                /* Make sure the whole structure fits in the table */
+                if ((unsigned long)(next - buf) > len)
```
here too

poettering commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,768 @@
+         * parse error.
+         */
+        if (read_full_file_full(AT_FDCWD, devmem, flags & FLAG_NO_FILE_OFFSET ? 0 : base, len, 0, NULL, (char **) &buf, &size) < 0) {
+                log_error("Failed to read table, sorry.");
```
always output errors, please:

```c
r = read_full_file_full(…);
if (r < 0) {
        return log_error_errno(r, "Failed to read table, sorry: %m");
```

poettering commented on src/udev/memory_id/types.h:
```diff
@@ -0,0 +1,63 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
```
i think we don't need this header, do we? it's only used by a single .c file after all. move everything in there?

poettering commented on src/udev/memory_id/dmidecode.c:
```diff
@@ -0,0 +1,768 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
```
i wonder if the helper should be called dmi_memory_id or so? memory_id sounds overly generic I think? dunno?

hadess replied:
Renamed to dmi_memory_id.

poettering commented:
Hmm, so what's the precise rationale behind this btw? It's a permission thing, right? i.e. DMI data in sysfs is not readable to unpriv clients, but the udev db is, thus the idea is to pull the relevant data out of sysfs and into udev, did I get this right?

hadess commented:
> Hmm, so what's the precise rationale behind this btw? It's a permission thing, right? i.e. DMI data in sysfs is not readable to unpriv clients, but the udev db is, thus the idea is to pull the relevant data out of sysfs and into udev, did I get this right?

That's exactly it. If you're happy with the general approach, I can certainly make the changes you requested in the review. I'll have to find an "indent" command-line that corresponds to your tastes ;)

hadess commented:
We discussed the rationale in https://github.com/systemd/systemd/issues/16651

poettering commented:
> We discussed the rationale in #16651

Ah, right, already had forgotten about htat.

maybe include a comment in the git commit msg about the rationale for this.

Hmm, and thinking about this more we should name the helper dmi_id, not memory_id or dmi_memory_id or so. For the simple reason that sooner or later we might want to export more info like that, that is not directly related to RAM. (e.g. battery info or bios language info or so) And thus having a name that leaves the door open for such future extensions is a good thing I think.

hadess commented:
Sure, I'll get on to the fixes soon, when I have a bit of spare time then.

hadess commented:
OK. there's only a few coding style changes questions in there. I'll squash all this together once you've had a look.

I'm sure there's more cleanups possible, as I removed about 6k lines of code from a 20-year old project that never modernised...

keszybz commented on src/udev/dmi_memory_id/dmi_memory_id.c:
```diff
@@ -0,0 +1,806 @@
+#define FLAG_NO_FILE_OFFSET     (1 << 0)
+#define FLAG_STOP_AT_EOT        (1 << 1)
+#define FLAG_FROM_DUMP          (1 << 2)
+static uint32_t opt_flags = 0;
```
Let's maybe simplify this to:
```C
static bool arg_no_file_offset = false;
static bool arg_stop_at_eot = false;
static const char *source_file = NULL;
```

keszybz commented on src/udev/dmi_memory_id/dmi_memory_id.c:
```diff
@@ -0,0 +1,806 @@
+        if (code >= 0x01 && code < ELEMENTSOF(technology))
+                printf("MEMORY_DEVICE_%u_MEMORY_TECHNOLOGY=%s\n", slot_num, technology[code]);
+        else
+                printf("MEMORY_DEVICE_%u_MEMORY_TECHNOLOGY=%s\n", slot_num, OUT_OF_SPEC_STR);
```
```C
printf("MEMORY_DEVICE_%u_MEMORY_TECHNOLOGY=%s\n", slot_num,
          code >= 0x01 && code < ELEMENTSOF(technology) ? technology[code] : OUT_OF_SPEC_STR);
```
?

keszybz commented on src/udev/dmi_memory_id/dmi_memory_id.c:
```diff
@@ -0,0 +1,806 @@
+        return 1;
+}
+
+int main(int argc, char * const argv[]) {
```
Please do
```C
static int run(int argc, const char** argv) {
       ...
}
DEFINE_MAIN_FUNC(run);
```
and use normal negative returns in `run`.

hadess replied:
``DEFINE_MAIN_FUNCTION(run);```

keszybz reviewed (commented):
I think we should do some more minor cleanups and merge this. The code works in general, and we can do various cleanups later on if we wish.

hadess commented on src/udev/dmi_memory_id/dmi_memory_id.c:
```diff
@@ -0,0 +1,806 @@
+                        printf("%s\n", GIT_VERSION);
+                        exit(EXIT_SUCCESS);
+                case '?':
+                        return -1;
```
None of the udev helpers do that, FWIW.

hadess replied:
It was such a win, that meant that the parsed options were now globals, and we were passing globals to the `_decode()` functions. I think it's a bit gross...

hadess commented on src/udev/dmi_memory_id/dmi_memory_id.c:
```diff
@@ -0,0 +1,806 @@
+        if (read_full_file_full(AT_FDCWD,
+                                opt_flags & FLAG_FROM_DUMP ? dump_file : SYS_ENTRY_FILE,
+                                0, 0x20, 0, NULL, (char **) &buf, &size) < 0) {
+                log_debug_errno(errno, "reading DMI from %s failed: %m",
```
```LOG_ERR```

hadess replied:
There was an argument missing in the `log_full_errno` example too.

hadess commented:
Let me know if you want me to squash this. I've also run script locally against a couple of dmi dumps I have, and checked that the output didn't change. Do we want to integrate those in the test suite?

keszybz commented:
> Let me know if you want me to squash this.

As long as there is one commit in the end, it doesn't matter. (We can easily squash when merging in the github interface). But when there's at least two commits desired, please squash into the final form.

> I've also run script locally against a couple of dmi dumps I have, and checked that the output didn't change. Do we want to integrate those in the test suite?

Yes, please add the test data. I don't know how familiar you are with meson: if you don't want to dive into this yourself, I can do the test framework integration bits.

hadess commented on src/udev/dmi_memory_id/dmi_memory_id.c:
```diff
@@ -0,0 +1,791 @@
+static void dmi_memory_device_type_detail(unsigned slot_num, uint16_t code) {
+        /* 7.18.3 */
+        static const char *detail[] = {
+                [1] = "Other",
```
The original code used base-10 indices, which I didn't transform.

hadess commented on src/udev/dmi_memory_id/dmi_memory_id.c:
```diff
@@ -0,0 +1,791 @@
+        }
+}
+
+static void dmi_table(off_t base, uint32_t len, uint16_t num, const char *devmem, bool stop_at_eot) {
```
`off_t` is signed though, `size_t` is the unsigned one.

hadess commented on src/udev/dmi_memory_id/dmi_memory_id.c:
```diff
@@ -0,0 +1,791 @@
+
+        for (a = 0; a < len; a++)
+                sum += buf[a];
+        return (sum == 0);
```
I prefer it being there, but I'll remove it if you prefer.

hadess commented on src/udev/dmi_memory_id/dmi_memory_id.c:
```diff
@@ -0,0 +1,791 @@
+                unsigned slot_num, ULargeInteger code) {
+        /* 7.18.12 */
+        /* 7.18.13 */
+        if ((code.h == 0xFFFFFFFF && code.l == 0xFFFFFFFF) ||
```
This is the documentation for that field:
![Screenshot from 2020-12-15 14-34-44](https://user-images.githubusercontent.com/769861/102221680-b9764900-3ee2-11eb-91f1-a651f6bcb53a.png)

hadess replied:
I hadn't actually change this code from the original code in dmidecode. I've added LU as a suffix now.

hadess commented:
> Yes, please add the test data. I don't know how familiar you are with meson: if you don't want to dive into this yourself, I can do the test framework integration bits.

I wrote something with my own dmidecode. I don't think the other one adds anything to the tests, and it wasn't from one of my own machines. Additional `*.bin` file can be created with `dmidecode --dump-bin <file.bin>`, and running `dmi_memory_id -F <file.bin>` to generate the expected output.

poettering commented on src/udev/dmi_memory_id/dmi_memory_id.c:
```diff
@@ -0,0 +1,808 @@
+ */
+#define WORD(x) (unaligned_read_le16(x))
+#define DWORD(x) (unaligned_read_le32(x))
+#define QWORD(x) (to_ularge_integer(DWORD(x), DWORD(x + 4)))
```
why retain uLargeInteger? Isn't that just a remnant from really old C compilers which didn't have 64bit integers?

this should just use unaligned_read_le64() here, no? anything special about this that requires other accessors?

poettering replied:
(if ULargeInteger has some value (which i don't see) then it should probably be marked with `_packet_`)

hadess commented:
All good? the only test failure is on s390 and not related to this patchset.

poettering commented on src/udev/dmi_memory_id/dmi_memory_id.c:
```diff
@@ -0,0 +1,808 @@
+                return log_full_errno(!arg_source_file && r == -ENOENT ? LOG_DEBUG : LOG_ERR,
+                                      r, "Reading \"%s\" failed: %m",
+                                      arg_source_file ?: SYS_ENTRY_FILE);
+        }
```
coding style: we usually don't place {} around single-line if blocks

hadess replied:
Is it a single-line block if it uses multiple lines? :)

hadess replied:
Made the change so I can mark this as fixed.

keszybz replied:
> Is it a single-line block if it uses multiple lines? :)

All that matters is that's it's a single *logical* line. The number of source lines doesn't matter ;)

poettering commented on src/udev/dmi_memory_id/dmi_memory_id.c:
```diff
@@ -0,0 +1,808 @@
+                no_file_offset = true;
+        }
+
+        if (size >= 24 && memcmp(buf, "_SM3_", 5) == 0) {
```
doesn't matter but we have memory_startswith() that would make the part right of the && a bit prettier:

```c
if (… && memory_startswith(buf, size, "_SM3_")) {
…
```

but doesn't matter

poettering commented on src/udev/dmi_memory_id/dmi_memory_id.c:
```diff
@@ -0,0 +1,766 @@
+        return 1;
+}
+
+static int smbios_decode(uint8_t *buf, const char *devmem, bool no_file_offset) {
```
hmm, so these functions, they effectively return booleans, but they are called "int". Which I guess is a C89-ism left-over.

And I think they follow the logic here that they return 0 on error, and 1 on success. 

Our usual coding style is different though, more kernel-like. Specifically, we always return negative errno-style errors on failure and >= 0 on success. We almost never do bool returns, except if the function explicitly is a boolean predicate thing (i.e. has no side effects, and tests something for yes and no, such as your verify_checksum() function).

poettering commented on src/udev/dmi_memory_id/dmi_memory_id.c:
```diff
@@ -0,0 +1,766 @@
+                log_error("Entry point length too large (%u bytes, expected %u).",
+                          (unsigned)buf[0x05], 0x1FU);
+                return 0;
+        }
```
continuing from above, following our usual coding style we would code this as:

```c
if (buf[0x05] > 0x20) 
        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "…);
```

i.e. propagate an roughly reasonably error code (we wrap it in a SYNTHETIC_ERRNO() macro call in this case, which we do whenver we generate the error here, instead of propagating some error code we received from some other call which generated the error itself

poettering commented on src/udev/dmi_memory_id/dmi_memory_id.c:
```diff
@@ -0,0 +1,766 @@
+static int smbios_decode(uint8_t *buf, const char *devmem, bool no_file_offset) {
+        /* Don't let checksum run beyond the buffer */
+        if (buf[0x05] > 0x20) {
+                log_error("Entry point length too large (%u bytes, expected %u).",
```
If we wanted to be fully correct:

`…%u…` →  `…%" PRIu8 "…`

(and then drop the `(unsigned)` cast)

poettering commented:
Code looks good to me. The error propagation thing doesn't matter too much, this new code is already at a higher standard then some of the other builtins, and it really doesn't matter much.

I'd be fine with merging it like this (of course an error propagation rework would be very welcome still).

Maybe @keszybz or @yuwata have an opinion on this before we merge it though?

I'll merge it in a few days if they don't report back or beat me to it

hadess commented:
> Code looks good to me. The error propagation thing doesn't matter too much, this new code is already at a higher standard then some of the other builtins, and it really doesn't matter much.

I only have a couple of workdays left before I leave for the holidays, so I'd really appreciate if we could land it as-is and assign me an issue to look into the error reporting. I would need to figure out how to create broken entries to exercise those, and I really don't want this to be something I have to scramble to correct for a release if I get it wrong, and certainly not during the holidays ;)

yuwata commented:
@hadess I added a commit which fixes several issues. PTAL.

hadess commented:
> @hadess I added a commit which fixes several issues. PTAL.

I'm sorry, but this is unreadable to me.

yuwata commented:
> There's trailing spaces in the text files. I couldn't even commit those files without making them pass off as binaries.

I made the trailing spaces dropped.

yuwata commented:
> > @hadess I added a commit which fixes several issues. PTAL.
> 
> I'm sorry, but this is unreadable to me.

Huh? You already commented on my commit. What do you mean?

hadess commented:
> Huh? You already commented on my commit. What do you mean?

That I can't parse 600 lines of diff without an explanation as to what the changes are supposed to do, trying to guess what you thought the problems were, and whether it was a good way to fix those problems.

yuwata commented on src/udev/dmi_memory_id/dmi_memory_id.c:
```diff
@@ -145,123 +130,136 @@ static void dmi_print_memory_size(
+                [0x07] = "CRC",
         };
 
-        if (code < ELEMENTSOF(type) && code[type])
```
`code[type]` -> `type[code]`

hadess replied:
I don't know how the compiler doesn't catch this... Good spot.

yuwata commented on src/udev/dmi_memory_id/dmi_memory_id.c:
```diff
@@ -145,123 +130,136 @@ static void dmi_print_memory_size(
+                [0x10] = "Die",
         };
 
-        if (code < ELEMENTSOF(form_factor) && code[form_factor])
```
`code[form_factor]` -> `form_factor[code]`

hadess replied:
Ditto.

yuwata commented on src/udev/dmi_memory_id/dmi_memory_id.c:
```diff
@@ -271,77 +269,75 @@ static void dmi_memory_device_set(unsigned slot_num, uint8_t code) {
+                [0x21] = "HBM2",
         };
 
-        if (code < ELEMENTSOF(type) && code[type])
```
`code[type]` -> `type[code]`

yuwata commented on src/udev/dmi_memory_id/dmi_memory_id.c:
```diff
@@ -351,43 +347,43 @@ static void dmi_memory_device_type_detail(unsigned slot_num, uint16_t code) {
+
         printf("MEMORY_DEVICE_%u_MEMORY_TECHNOLOGY=%s\n", slot_num,
-                        code < ELEMENTSOF(technology) && code[technology] ? technology[code] : OUT_OF_SPEC_STR);
+               code < ELEMENTSOF(technology) && technology[code] ? technology[code] : OUT_OF_SPEC_STR);
```
`code[technorlogy]` -> `technology[code]`

yuwata commented on src/udev/dmi_memory_id/dmi_memory_id.c:
```diff
@@ -471,107 +457,96 @@ static void dmi_decode(const struct dmi_header *h) {
-static void dmi_table_decode(uint8_t *buf, uint32_t len, uint16_t num, bool stop_at_eot) {
-        uint8_t *data;
-        int i = 0;
+static void dmi_table_decode(const uint8_t *buf, size_t len, uint16_t num) {
```
constify `buf`.

yuwata commented on src/udev/dmi_memory_id/dmi_memory_id.c:
```diff
@@ -584,24 +559,17 @@ static void dmi_table_decode(uint8_t *buf, uint32_t len, uint16_t num, bool stop
                 data = next;
-
-                /* SMBIOS v3 requires stopping at this marker */
-                if (h.type == 127 && stop_at_eot)
```
The condition is redundant. Hence, I dropped.

yuwata commented on src/udev/dmi_memory_id/dmi_memory_id.c:
```diff
@@ -145,123 +130,136 @@ static void dmi_print_memory_size(
+        char *str;
+
+        str = strdupa(dmi_string(h, s));
+        printf("MEMORY_DEVICE_%u_%s=%s\n", slot_num, attr_suffix, strstrip(str));
```
trailing white space is now dropped.

yuwata reviewed (commented):
> That I can't parse 600 lines of diff without an explanation as to what the changes are supposed to do, trying to guess what you thought the problems were, and whether it was a good way to fix those problems.

@hadess Uh, oh, I see. Sorry for my big commit. I briefly explained what I made. See the below comments. Other changes are mostly cosmetic.

yuwata commented:
@hadess BTW, if you find anything bad in my commits, then feel free to drop my commits or partially revert my change. These commits are just my suggestions.

hadess commented on src/udev/dmi_memory_id/dmi_memory_id.c:
```diff
@@ -145,123 +130,136 @@ static void dmi_print_memory_size(
-        if (code < ELEMENTSOF(type) && code[type])
-                return type[code];
-        return OUT_OF_SPEC_STR;
+        if (code != 0x03) /* Do not print "None". */
```
Why? Wouldn't it be better to know that the property existed and was set, rather than not print it?

yuwata replied:
You dropped that value previously. Now, `None` is also shown.

hadess replied:
Oh, I got this wrong then, sorry.

hadess commented on src/udev/dmi_memory_id/dmi_memory_id.c:
```diff
@@ -145,123 +130,136 @@ static void dmi_print_memory_size(
-                printf("MEMORY_DEVICE_%u_PRESENT=0\n", slot_num);
-                return;
-        } else if (code == 0xFFFF) {
+        uint64_t s = code & 0x7FFF;
```
I would only have initialised it before being used.

yuwata replied:
OK.

hadess commented on src/udev/dmi_memory_id/dmi_memory_id.c:
```diff
@@ -444,21 +439,12 @@ static void dmi_decode(const struct dmi_header *h) {
-                }
+                dmi_memory_array_location(data[0x04]);
+                dmi_memory_array_ec_type(data[0x06]);
+                if (DWORD(data + 0x07) != 0x80000000)
```
This doesn't look like the correct simplification of those conditionals.

keszybz replied:
I unset the flag since we probably need to clarify that before merging.

yuwata replied:
Sorry, I cannot find any error on this change. Could you elaborate more?

keszybz replied:
`data[0x06] == 0x03` will now print MEMORY_ARRAY_EC_TYPE=None, but before it didn't.

yuwata replied:
I made `dmi_memory_array_ec_type()`  not show `None`. But in the above comment, @hadess said 'Why?' So, now the EC_TYPE= is always shown.

hadess replied:
> I made `dmi_memory_array_ec_type()` not show `None`. But in the above comment, @hadess said 'Why?' So, now the EC_TYPE= is always shown.

Not sure why you're commenting about the EC Type changes, when the problem is the conditional the line below.

hadess replied:
But I didn't see the fact that the conditional got reversed. That's why I said that I couldn't review the monster diff of patches :/

yuwata commented:
@hadess BTW, should we expose `PartNumber=` even if it is empty?
```
# dmidecode
(snip)
Handle 0x0004, DMI type 17, 40 bytes
Memory Device
        Array Handle: 0x0003
        Error Information Handle: Not Provided
        Total Width: 64 bits
        Data Width: 64 bits
        Size: 4 GB
        Form Factor: SODIMM
        Set: None
        Locator: ChannelA-DIMM0
        Bank Locator: BANK 0
        Type: DDR4
        Type Detail: Synchronous Unbuffered (Unregistered)
        Speed: 2400 MT/s
        Manufacturer: 0000
        Serial Number: 00000000
        Asset Tag: None
        Part Number:                     
        Rank: 1
        Configured Memory Speed: 2400 MT/s
        Minimum Voltage: Unknown
        Maximum Voltage: Unknown
        Configured Voltage: 1.2 V
(snip)
```
```
# dmi_memory_id | grep MEMORY_DEVICE_0
MEMORY_DEVICE_0_TOTAL_WIDTH=64
MEMORY_DEVICE_0_DATA_WIDTH=64
MEMORY_DEVICE_0_SIZE=4294967296
MEMORY_DEVICE_0_FORM_FACTOR=SODIMM
MEMORY_DEVICE_0_LOCATOR=ChannelA-DIMM0
MEMORY_DEVICE_0_BANK_LOCATOR=BANK 0
MEMORY_DEVICE_0_TYPE=DDR4
MEMORY_DEVICE_0_TYPE_DETAIL=Synchronous Unbuffered (Unregistered)
MEMORY_DEVICE_0_SPEED_MTS=2400
MEMORY_DEVICE_0_MANUFACTURER=0000
MEMORY_DEVICE_0_SERIAL_NUMBER=00000000
MEMORY_DEVICE_0_ASSET_TAG=None
MEMORY_DEVICE_0_PART_NUMBER=
MEMORY_DEVICE_0_RANK=1
MEMORY_DEVICE_0_CONFIGURED_SPEED_MTS=2400
MEMORY_DEVICE_0_CONFIGURED_VOLTAGE=1
```

yuwata commented:
Hmm, consistently fails to build on s390x...
```
FAILED: src/udev/dmi_memory_id 
cc  -o src/udev/dmi_memory_id 'src/udev/3eb56e7@@dmi_memory_id@exe/dmi_memory_id_dmi_memory_id.c.o' -flto -Wl,--as-needed -Wl,--no-undefined -pie -Wl,-z,relro -Wl,-z,now -fstack-protector -Wl,--gc-sections -g -O2 -fdebug-prefix-map=/tmp/autopkgtest.0fdnM9/build.yd1/systemd=. -fstack-protector-strong -Wformat -Werror=format-security -Wl,-Bsymbolic-functions -Wl,-z,relro -Wl,--start-group src/shared/libsystemd-shared-247.a src/libsystemd/libsystemd_static.a src/basic/libbasic.a -pthread -lacl /usr/lib/s390x-linux-gnu/libblkid.so /usr/lib/s390x-linux-gnu/libcap.so -lcrypt -lgcrypt /usr/lib/s390x-linux-gnu/libip4tc.so /usr/lib/s390x-linux-gnu/libip6tc.so /usr/lib/s390x-linux-gnu/libkmod.so /usr/lib/gcc/s390x-linux-gnu/7/../../../s390x-linux-gnu/liblz4.so /usr/lib/s390x-linux-gnu/libmount.so -lpam -lrt /usr/lib/s390x-linux-gnu/libseccomp.so /usr/lib/s390x-linux-gnu/libselinux.so /usr/lib/gcc/s390x-linux-gnu/7/../../../s390x-linux-gnu/libzstd.so /usr/lib/s390x-linux-gnu/liblzma.so -lm -ldl -Wl,--end-group '-Wl,-rpath,$ORIGIN/../shared:$ORIGIN/../libsystemd:$ORIGIN/../basic' -Wl,-rpath-link,/tmp/autopkgtest.0fdnM9/build.yd1/systemd/build-deb/src/shared -Wl,-rpath-link,/tmp/autopkgtest.0fdnM9/build.yd1/systemd/build-deb/src/libsystemd -Wl,-rpath-link,/tmp/autopkgtest.0fdnM9/build.yd1/systemd/build-deb/src/basic
/usr/lib/s390x-linux-gnu/libselinux.so: undefined reference to `pthread_once'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
dh_auto_build: error: cd build-deb && LC_ALL=C.UTF-8 ninja -j1 -v returned exit code 1
debian/rules:189: recipe for target 'override_dh_auto_build' failed
```

hadess commented:
> @hadess BTW, should we expose `MEMORY_DEVICE_X_PART_NUMBER=` even if it is empty?

In this case I would, that means that an empty string is set as the part number, rather than the part number not being filled in. My RAM modules all have part numbers here...

hadess commented:
I've reverted the printing of EC type, and fixed the output from the test in case of errors (`echo $foo` will remove linefeeds, but `echo "$foo"` will show them, learnt something).

Ready to merge as far as I'm concerned.

hadess commented:
> Hmm, consistently fails to build on s390x...

Looks like the SELinux library is underlinked.

yuwata commented:
LGTM, except for the build failure on s390x.

keszybz commented:
I think we should ignore the s390x failure for now. Let's report the issue downstream. If it doesn't get fixed, we'll try some workaround.

hadess commented:
> I think we should ignore the s390x failure for now. Let's report the issue downstream. If it doesn't get fixed, we'll try some workaround. (Or is dmi even useful on s390x? We could skip it if not.)

It isn't, but the code should still compile and run (against binary dumps) just fine.

keszybz commented:
> It isn't, but the code should still compile and run (against binary dumps) just fine.

Sure, but we probably shouldn't install a udev plugin that cannot have any effect. It's one thing to build something for testing, and another to include it in the default install.

keszybz commented:
bionic-amd64 failed in "boot-smoke", hopefully unrelated.

hadess commented:
> Sure, but we probably shouldn't install a udev plugin that cannot have any effect. It's one thing to build something for testing, and another to include it in the default install.

https://github.com/systemd/systemd/pull/18001

poettering commented:
hmm, the dmidecode dumps were never commited, and Lennart can't build/test systemd anymore... See #18148 

(not sure why i didn't notice this before...)
